### PR TITLE
feat: Collection Component

### LINKED ISSUE #987: New Component: Collection
# [Collection](https://designsystem.digital.gov/components/collection/)

## USWDS Version

2.10.0

## Description

A collection displays a compact list of multiple related items like articles or events. The list links each item to its original source.

## Requirements (proposed)

**Prop Interface**

```JSX
interface CollectionMetaItemProps {
  text: string
  className?: string
}

interface CollectionMetaTagProps {
  text: string
  style?: "default" | "new" //default "default"
  className?: string
}

interface BaseCollectionItemProps {
  heading: string
  href: string
  className?: string
  liProps?: JSX.IntrinsicElements['li']
}

interface CollectionItemProps extends BaseCollectionItemProps {
  description: string
  metaTags: CollectionMetaTagProps[]
}

interface MediaCollectionItemProps extends CollectionItemProps {
  src: string
  alt: string
}

interface CalendarCollectionItems extends CollectionItemProps {
  datetime: string 
  // Or maybe month: string, day: string
  // Either way, one should programmatically determine the other for JSX markup
}

interface HeaderOnlyCollectionItems extends BaseCollectionItemProps {
  metaItems: CollectionMetaItemProps[]
}


interface CollectionProps {
  children: React.ReactElement<BaseCollectionItemProps> | React.ReactElement<BaseCollectionItemProps>[]
  condensed?: boolean //default false. This should be used when the collection items are HeaderOnlyCollectionItems. Checking the type of children at runtime may also work, except when a user mix-n-matches collection items. More thought necessary here
  className?: string
  listProps?: JSX.IntrinsicElements['ul']
}
```

## Example Implementation (proposed)

```JSX
  <Collection>
    <CollectionItem 
      heading="Gears of Government President’s Award winners" 
      href="https://www.performance.gov/presidents-winners-press-release/" 
      description ="Today, the Administration announces the winners of the Gears of Government President’s Award. This program recognizes the contributions of individuals and teams across the federal workforce who make a profound difference in the lives of the American people." 
      metaTags={[
        {text: "New", style: "new"}, // Could also implement as <CollectionMetaTag text="New" style="new"/>. Internally breaking these out into components might be nice.
        {text: "PMA", style: "default"}, 
        {text: "OMB", style: "default"}
      ]}
    />
    ...
  </Collection>
```
```JSX
  <Collection>
    <ThumbnailCollectionItem 
      heading="Gears of Government President’s Award winners" 
      href="https://www.performance.gov/presidents-winners-press-release/"
      src="https://www.performance.gov/img/GoG/gears-govt-presidents.png"
      alt="Alt Text"
      description ="Today, the Administration announces the winners of the Gears of Government President’s Award. This program recognizes the contributions of individuals and teams across the federal workforce who make a profound difference in the lives of the American people." 
      metaTags={[
        {text: "New", style: "new"}, 
        {text: "PMA", style: "default"}, 
        {text: "OMB", style: "default"}
      ]}
    />
  ...
  </Collection>
```
```JSX
  <Collection>
    <CalendarCollectionItem 
      heading="Gears of Government President’s Award winners" 
      href="https://www.performance.gov/presidents-winners-press-release/" 
      datetime="2020-09-30T12:00:00+01:00"
      description ="Today, the Administration announces the winners of the Gears of Government President’s Award. This program recognizes the contributions of individuals and teams across the federal workforce who make a profound difference in the lives of the American people." 
      metaTags={[
        {text: "New", style: "new"}, // Could also implement as <CollectionMetaTag text="New" style="new"/>. Internally breaking these out into components might be nice.
        {text: "PMA", style: "default"}, 
        {text: "OMB", style: "default"}
      ]}
    />
    ...
  </Collection>
```
```JSX
  <Collection condensed>
    <HeaderOnlyCollectionItem 
      heading="The eight principles of mobile-friendliness" 
      href="https://digital.gov/guides/mobile-principles/?dg" 
      metaItems={[
        { text: "Digital.gov" } // Could also implement as <CollectionMetaItem text="New" />. Internally breaking these out into components might be nice.
      ]}
    />
    ...
  </Collection>
```

**More Details**

- `<ul>` root element for Collection
- `<li>` root element for CollectionItems

## Other Considerations

- The different "types" of collection items are different _enough_ in terms of props and markup that I would recommend having separate functional components exposed for each (CollectionItem, CalendarCollectionItem, ThumbnailCollectionItem, HeadingCollectionItem)
- Since a good portion of the markup is, however, common between each type of CollectionItem, internally I would recommend making use of a "CollectionItemBody" or similar which a common set of props would be sufficient to render, and simply pass those through from each wrapping CollectionItem variant.
- For other components we typically use a "variant" prop to render different variants using the same component. In this case, that would feel a bit clumsy as each variant also have their own set of props in addition to the common props for the component body.
- The above are not rigid requirements. Spending time in development with this component may reveal better ways to implement it. These are just some recommendations at a first pass of the uswds component code.

### PR DESCRIPTION
**Pull request by christopherhuii**
# Summary
# [Collection](https://designsystem.digital.gov/components/collection/)
This PR adds the USWDS Collection component. The` Collection` is a list of `CollectionItem`s. `CollectionItem`s are composed of smaller building blocks (subcomponents). They can have:
1. `CollectionHeading`
2. `CollectionDescription`
3. `CollectionMeta`
4. `CollectionMetaItem`
5. `CollectionMetaItemTag`
6. `CollectionThumbnail`
7. `CollectionCalendarDate`

Each of these subcomponents are used in the stories in Storybook. I have built out the examples on [USWDS](https://designsystem.digital.gov/components/collection/), but developers are able to add/remove pieces as needed.

Closes #987

## How To Test
This component isn't dynamic. It's mainly just markup and styling. Verifying code and a visual check in storybook should be enough.

### Note
On USWDS, there is a "headings only" version of the collection item. I've built that, WITHOUT the globe SVG icon. That icon is going to be from the [Icon](https://designsystem.digital.gov/components/icon/) component so I omitted it for now. Once the `Icon` component is implemented, all we'd need to do is update the storybook stories to reflect its usage.

### CODE REVIEW

**Comment by trussworks-infra** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true">Warnings</th>
    </tr>
  </thead>
  <tbody><tr>
      <td>:warning:</td>
      <td>This PR does not include changes to storybook, even though it affects component code.</td>
    </tr>
  
<tr>
      <td>:warning:</td>
      <td>It looks like there are new component (JSX/TSX) files, but the entrypoint (index.ts) has not changed. - <em>Did you forget to export new components from the library entrypoint?</em></td>
    </tr>
  
<tr>
      <td>:warning:</td>
      <td>This PR does not include changes to tests, even though it affects source code.</td>
    </tr>
  </tbody>
</table>



<p align="right">
  Generated by :no_entry_sign: <a href="https://danger.systems/js">dangerJS</a> against a78a3b20bb2bb36f695f343f1da830003bb77818
</p>

**Review by suzubara — COMMENTED** [review]
Just some high level notes, since I think the recommended approach has changed since the original issue was written. We would like to lean more towards using subcomponents to compose different variations, rather than pass in content via multiple layers of props.

possible components might be:
- `<Collection>` - the root `ul`
- `<CollectionItem>` - each `li` element, these could take props to show the calendar date
- `<CollectionHeading>, <CollectionDescription>, <CollectionMeta>, <CollectionMetaItem>` - contents of an item
- `CollectionImage, CollectionCalendarDate` - image, calendar elements

example implementations could look like:
```
<Collection>
  <CollectionItem>
    <CollectionHeading><a className="usa-link">Here is a link</a></CollectionHeading>
    <CollectionDescription>Today, we published...</CollectionDescription>
    <CollectionMeta aria-label="More information">
      <CollectionMetaItem>By Sonda Ainsworth and Constance Lu</CollectionMetaItem>
    </CollectionMeta>
    <CollectionMeta aria-label="Topics">
      <CollectionMetaItem tag tagStyle="new">New</CollectionMetaItem>
      <CollectionMetaItem tag>PMA</CollectionMetaItem>
    </CollectionMeta>
  </CollectionItem>

  <CollectionItem image={<CollectionImage src="..." />}>
    <CollectionHeading><a className="usa-link">Here is a link</a></CollectionHeading>
    <CollectionDescription>Today, we published...</CollectionDescription>
    <CollectionMeta aria-label="More information">
      <CollectionMetaItem>By Sonda Ainsworth and Constance Lu</CollectionMetaItem>
    </CollectionMeta>
    <CollectionMeta aria-label="Topics">
      <CollectionMetaItem tag tagStyle="new">New</CollectionMetaItem>
      <CollectionMetaItem tag>PMA</CollectionMetaItem>
    </CollectionMeta>
  </CollectionItem>

  <CollectionItem calendar={{ datetime: '2020-09-30T12:00:00+01:00', href: '...' }}>
    <CollectionHeading><a className="usa-link">Here is a link</a></CollectionHeading>
    <CollectionDescription>Today, we published...</CollectionDescription>
    <CollectionMeta>
      <CollectionMetaItem>By Sonda Ainsworth and Constance Lu</CollectionMetaItem>
    </CollectionMeta>
    <CollectionMeta aria-label="Topics">
      <CollectionMetaItemTag className="usa-tag--new">New</CollectionMetaItemTag>
      <CollectionMetaItemTag>PMA</CollectionMetaItemTag>
    </CollectionMeta>
  </CollectionItem>
</Collection>
```

File: src/components/Collection/Collection.stories.tsx [hunk 1]
```diff
@@ -0,0 +1,510 @@
...
+  },
+}
+
+export const collectionWithDefaultItems = (): React.ReactElement => (
+  <div className="tablet:grid-col-6">
+    <Collection>
+      <CollectionItem>
+        <CollectionHeading>
+          <a
+            className="usa-link"
```
Comment (brandonlenz) [inline src/components/Collection/Collection.stories.tsx:32]: You can replace anchor tags with class `usa-link` with the `Link` component we have internally

File: src/components/Collection/Collection.stories.tsx [hunk 2]
```diff
@@ -0,0 +1,510 @@
...
+          who make a profound difference in the lives of the American people.
+        </CollectionDescription>
+        <CollectionMeta aria-label="More Information">
+          <CollectionMetaItem>
+            By Sondra Alnsworth and Constance Lu
+          </CollectionMetaItem>
+          <CollectionMetaItem>September 30, 2020</CollectionMetaItem>
+        </CollectionMeta>
+        <CollectionMeta aria-label="Topics">
+          <CollectionMetaItemTag className="usa-tag--new">
```
Comment (brandonlenz) [inline src/components/Collection/Collection.stories.tsx:50]: I always look for ways to encapsulate as many of the USWDS classes as possible. What do you think about Suz's suggestion of a `tagStyle` prop where the value `"new"` can apply this class internally?

**Review by brandonlenz — COMMENTED** [review]
A few small comments. I think this looks amazing!

**Comment by christopherhuii** [comment]
@brandonlenz Thanks for the feedback! I made the following changes:
- Add `tagStyle` prop (`'new' | 'default'`) in `CollectionMetaItemTag`
- Update stories accordingly with `Link` and `tagStyle`
- Pull latest changes from `main`

**Review by brandonlenz — APPROVED** [review]
Looks great! Thanks for making those changes

**Pull request merged**